### PR TITLE
feat(plugin): current function symbol

### PR DESCRIPTION
**Pull request by pyrho**
tl;dr
====
* New plugin action: `getCurrentFunctionSymbol`
* New preference: `coc.preferences.currentFunctionSymbolAutoUpdate`
* New depedency on `binary-search`.

---

A new handler method had been added (`getCurrentFunctionSymbol`) that
returns the name of the enclosing
Class/Method/Function/Interface/Enum/Constructor relative to the current
cursor position.
It will also set the `b:coc_current_function` buffer variable.

If the user choses to, by setting the
`coc.preferences.currentFunctionSymbolAutoUpdate` to `true` (default:
false), a hook is setup on the `CursorHold` event that will call
`getCurrentFunctionSymbol` and subsequently update the value of
`b:coc_current_function`.
This is useful in the context of an enhanced status line for instance.

Fixes https://github.com/neoclide/coc.nvim/issues/512

Inspired by @liuchengxu 's work on https://github.com/liuchengxu/vista.vim .

### CODE REVIEW

**Comment by pyrho** [comment]
Example lightline integration:
```
function! CocFunc()
    let currentFunctionSymbol = get(b:, 'coc_current_function', '')
    return currentFunctionSymbol !=# '' ? "\uf6a6 " .currentFunctionSymbol : ''
endfunction

let g:lightline.component_function = {
            \ 'cocstatus': 'coc#status',
            \ 'currentFunc': 'CocFunc'
            \}
let g:lightline.active = {
            \ 'left': [ [ 'mode', 'paste', ],
            \           [ 'readonly', 'relativepath', 'modified' ],
            \           [ 'cocstatus', 'currentFunc' ] ],
           ....
           \}
```

Screenshot 2019-04-08 at 23 00 25

**Comment by pyrho** [comment]
I didn't add any test because I was not sure on how to proceed, how would I test this (if doable) ?

File: src/handler/index.ts [hunk 1]
```diff
@@ -195,6 +202,47 @@ export default class Handler {
     this.colors = new Colors(nvim)
   }
 
+  public async getCurrentFunctionSymbol(): Promise<string> {
+    let { position } = await workspace.getCurrentState()
+    const buffer = await this.nvim.buffer
+    let symbols = await this.getDocumentSymbols()
+
+    if (!symbols || symbols.length === 0) {
+        buffer.setVar('coc_current_function', '')
```
Comment (chemzqm) [inline src/handler/index.ts]: Should be `buffer.setVar('coc_current_function', '', true)`

File: src/handler/index.ts [hunk 2]
```diff
@@ -195,6 +202,47 @@ export default class Handler {
...
+
+    const currentFunctionName = (() => {
+      const sym = symbols[symbolPosition]
+      if (!sym || (position.line > sym.range.end.line)) {
+        return ''
+      } else {
+          return sym.text
+      }
+    })()
+    buffer.setVar('coc_current_function', currentFunctionName)
```
Comment (chemzqm) [inline src/handler/index.ts]: Should be `buffer.setVar('coc_current_function', currentFunctionName, true)`

File: src/handler/index.ts (hunk 1 above)
Comment (pyrho) [reply src/handler/index.ts]: Done.

File: src/handler/index.ts (hunk 2 above)
Comment (pyrho) [reply src/handler/index.ts]: Done.

File: src/handler/index.ts [hunk 3]
```diff
@@ -177,6 +179,11 @@ export default class Handler {
         nvim.command('pclose', true)
       }
     }, 100), null, this.disposables)
+    if (this.preferences.currentFunctionSymbolAutoUpdate) {
+      events.on('CursorHold', () => {
+        this.getCurrentFunctionSymbol()
```
Comment (chemzqm) [inline src/handler/index.ts]: Need change to `async  () => { await this.getCurrentFunctionSymbol()`, it will catch possible errors.

Comment (pyrho) [reply src/handler/index.ts]: ✔️
